### ISSUE TITLE
duffle build followed by duffle export should work

### ISSUE THREAD

**Issue by itowlson** [deislabs/duffle#828]
If you do a `duffle build` followed by a (full-bundle) `duffle export` then you get a baffling error (per #824), the significance of which is "you need to push the invocation image first," even though you may already have the invocation image in your local Docker registry as a result of `duffle build`.

The reason that Duffle always attempts an image pull is so that it can work in environments where there is no local Docker daemon (thanks @glyn for the explanation).  But this makes for a non-smooth build-then-export experience, not made easier by the cryptic error.

Some possible solutions:

1. `duffle build` could always push the invocation image.  @glyn has pointed out that this avoids problems where the user does a `duffle build` followed by `duffle export -t` (manifest only bundle).  At the moment, the exported thin bundle will Work On My Machine TM, but fail when I give it to someone else because I haven't published the image.  Downside is that maybe I don't want to put my invocation image up in public just yet; it also makes it harder to test locally, e.g. using Kind - these problems could be mitigated by a `duffle build --push-image=false` option.

2. `duffle export` could check whether there is a local Docker daemon, and if so check that before resorting to a pull.

Or something else of course!

**Comment by itowlson** [comment]
Another idea: `duffle export -t` could perform the push - this would address the 'exported thin bundle fails on consumers' machines' problem, and makes some kind of sense because `export` implies an intent to distribute/publish more than `build` does.

This would still require `duffle export` without the `-t` to check the local Docker daemon if there was one, as conceptually there's no need to push in the full bundle scenario - your images are included _in_ the export and don't need to be placed in a separate registry.

**Comment by glyn** [comment]
> ... these problems could be mitigated by a duffle build --push-image=false option

Agreed, it's not always appropriate for `duffle build` to push the image.

> duffle export could check whether there is a local Docker daemon, and if so check that before resorting to a pull.

That feels like a clean solution, but it's non-trivial. We would need to extend pivotal/image-relocation to support that (or, if that's not feasible, see if there is a different library for writing the OCI image layout).

> Another idea: `duffle export -t` could perform the push

I think that behaviour would be rather surprising. Also, the user can issue `duffle build -o` to get a thin bundle without using `duffle export -t` and then we are back to the same issue - if they give that thin bundle to someone using a different docker daemon, the other person can fall foul of the missing invocation image.

**Comment by glyn** [comment]
> duffle export could check whether there is a local Docker daemon, and if so check that before resorting to a pull.
> 
> That feels like a clean solution, but it's non-trivial. We would need to extend pivotal/image-relocation to support that (or, if that's not feasible, see if there is a different library for writing the OCI image layout).

Looking into this solution.

**Comment by glyn** [comment]
@itowlson Would you care to review https://github.com/pivotal/image-relocation/pull/18?

**Comment by itowlson** [comment]
@glyn I don't feel qualified to review it at a code level, but the approach looks good and I agree with your comment about error reporting.  Thanks!

**Comment by glyn** [comment]
It turns out that the option of exporting images from the docker daemon is unsound because the docker daemon makes no guarantee to create the same content digest for an image as the image's repository (which belongs to a specific registry with specific compression algorithms etc.). So we are back to the drawing board with this issue. The remaining options appear to be:
* `duffle build` providing an option, possibly even the default, to push the invocation image
* `duffle export` to push all images to ensure that some updates are not "stuck" in the daemon.

**Comment by radu-matei** [comment]
I think a combination of 1 and 3:

- add a flag to build to allow users to opt out of pushing by default
- export to push all of them anyway.

**Comment by glyn** [comment]
Agreed. Compare [cnab-to-oci](https://github.com/docker/cnab-to-oci) behaviour, which is analogous to export:
> Note: cnab-to-oci does not push images from your docker daemon image store. Make sure all your invocation images are already present on a registry before pushing your bundle.

**Comment by glyn** [comment]
Unassigning so someone else can take this. I won't be able to get round to it for some time.

**Comment by squarism** [comment]
Sorry to be naive here but why are layers coupled to a registry store?  If I don't have a registry, I have an image (made up of layers which are SHA'd).  Layers have to be transportable and cryptographically the same (otherwise, what's the point of docker?).  Is there something else that is coupling an image that isn't even pushed to a registry?

Let's say I pull down hello-world and tag is without a registry prefix.  Now I duffle up hello-world.  Can't duffle just write the tars from the image I have?  Easier said than done I know.  🌻

**Comment by glyn** [comment]
@squarism maybe a bit of background will help. Each image has two digests. The so-called _repository digest_ depends on the compressed form of the image and compression can vary by registry. Then there is an _image id_ which is a digest of the uncompressed form of the image and so is registry independent. If you include a digest in an image reference, it's a repository digest.

Some tools don't preserve the repo digest. For instance, if you docker pull/tag/push, the repo digest can change even though the uncompressed image stays the same. (Tools based on https://github.com/google/go-containerregistry, which is ultimately what `duffle relocate` uses, preserve the repo digest when copying images between registries.)
